aepsych/database/tables.py

```diff
@@ -368,11 +371,91 @@ def __repr__(self):
+                WHERE message_type = 'tell' AND master_table_id = {master_id}"
+                )
+                .fetchall()
+            )
+
+            # Get raw tab
+            for i, example in enumerate(tell_messages):
+                timestamp = example["timestamp"]
+                master_table_id = example["master_table_id"]
+
+                # Deserialize pickle message
+                message_contents = dill.loads(example["message_contents"])
+
+                # Get outcome
+                outcomes = message_contents["message"]["outcome"]
+                # Get parameters
+                params = message_contents["message"]["config"]
+                # Get model_data
+                model_data = message_contents["message"].get("model_data", True)
+
+                db_raw_record = server.record_raw(
+                    master_table=master_table,
+                    model_data=bool(model_data),
+                    timestamp=timestamp,
+                )
+
+                for param_name, param_value in params.items():
+                    if type(param_value) not in [float, int, bool]:
+                        if len(param_value) == 1:
+                            server.record_param(
+                                raw_table=db_raw_record,
+                                param_name=str(param_name),
+                                param_value=float(param_value[0]),
+                            )
+                        else:
+                            for j, v in enumerate(param_value):
+                                server.record_param(
+                                    raw_table=db_raw_record,
+                                    param_name=str(param_name) + "_stimuli" + str(j),
+                                    param_value=float(v),
+                                )
+                    else:
+                        server.record_param(
+                            raw_table=db_raw_record,
+                            param_name=str(param_name),
+                            param_value=float(param_value),
+                        )
+
+                if type(outcomes) not in [float, int, bool]:
+                    for j, outcome_value in enumerate(outcomes):
+                        server.record_outcome(
+                            raw_table=db_raw_record,
+                            outcome_name="outcome_" + str(j),
+                            outcome_value=float(outcome_value),
+                        )
+                else:
+                    server.record_outcome(
+                        raw_table=db_raw_record,
+                        outcome_name="outcome",
+                        outcome_value=float(outcomes),
+                    )
+
     @staticmethod
     def requires_update(engine):
+        """Check if the raw table is empty, and data already exists."""
+        n_raws = engine.execute("SELECT COUNT (*) FROM raw_data").fetchone()[0]
```

Comment:
What if `raw_data` table doesn't exist? Is it guaranteed to exist by the time this method is called?

Reply:
Hi!
I think it's guaranteed that the `raw_data` table exists when this method is called. If you have an old database, when you update your AEPsych version, the DB structure changes (defined in the `tables.py` file), so you will get an empty `raw_data` table 🤔

If you read an old DB with pure SQLAlchemy (without AEPsych), then the `raw_data` table won't exist, but in that case, this method won't be used.